src/spec/cli.cr

```diff
@@ -0,0 +1,148 @@
+require "./formatter"
+require "option_parser"
+
+module Spec
+  # :nodoc:
+  class CLI
+    # :nodoc:
+    class Options
+      property formatters = Array(Spec::Formatter).new
+      property locations = Array({String, Int32}).new
+
+      property? default_formatter : Spec::Formatter? = nil
```

Comment:
Lil' bit of nitpicking:
- No need for nilable postfix `?`, since it's added automatically by the `property?` macro
- Default `nil` assignments are redundant
- All properties except `Bool` IMHO should use normal `property` macro

ditto all below

Reply:
> No need for nilable postfix ?, since it's added automatically by the property? macro

I'm not sure that statement is correct. When `property?` macro expands, does not define a union:

```crystal
class Foo
  property? something : Bool
end
```

```
$ ccr tool expand -c 1.cr:2:3 1.cr 
Using compiled compiler at `.build/crystal'
Error in 1.cr:2: expanding macro

  property? something : Bool
  ^

in macro 'property?' expanded macro: macro_94853763181520:797, line 3:

   1.       
   2.         
>  3.           @something : Bool
   4. 
   5.           def something? : Bool
   6.             @something
   7.           end
   8. 
   9.           def something=(@something : Bool)
  10.           end
  11.         
  12.       
  13.     

instance variable '@something' of Foo was not initialized directly in all of the 'initialize' methods, rendering it nilable. Indirect initialization is not supported.
```

> Default nil assignments are redundant

Agree on that since `Type?` will default to `nil`. Will remove

> All properties except Bool IMHO should use normal property macro

I don't think so. Usage of `?`-style for the properties is not about boolean value, but the Crystal pattern to check for `nil`. This follows the pattern of `Hash#[]?`, `fetch?`, `first?` and many others.

Perhaps you refer to something else? Please explain if possible with a concrete example so I can better understand your comment.

Thank you.

Reply:
> I'm not sure that statement is correct. When property? macro expands, does not define a union:

My bad, it's the case only for `property/getter!` macros.

> I don't think so. Usage of ?-style for the properties is not about boolean value, but the Crystal pattern to check for nil. This follows the pattern of Hash#[]?, fetch?, first? and many others.

`?`-style getters are mostly used alongside of their `?`-less counterparts. Normal attributes tends to be defined as either of those, with `Exception#message` which return type is `String?` is one (counter-)example that comes to my mind.

Reply:
Thank you for your comments. I guess will wait for input from Crystal core in relation to the style used, as this is not outlined in Conventions / Coding style documentation.

Reply:
I'd say it is quite clear: Indicating a nilable property type with `?`-suffix is not common practice. A method with `?`-suffix makes only sense as an alternative to a method which would raise instead of returning `nil`. Another use case are *predicate methods* returning `Bool`.

All uses of the `property?`  macro in the standard library are with a `Bool` type. Nilable properties always use `property`.

Reply:
Good point, didn't explore the entire codebase back then to extrapolate that pattern, but I have done that last night.

Will still wait for Crystal core feedback prior sending further commits, as I have quite limited OSS time this week.

Cheers.

Reply:
`?` suffix is typically either asking a question or a nillable version of a method without a `?`. I wouldn't write it like this.

Reply:
> I wouldn't write it like this.

Can you be more specific? You wouldn't write it like shown in current code? you wouldn't change it to be non-`?`, or by the given examples in this conversation?

Please note that not everybody here is a native English speaker. Sometimes English is not even our second language. The more precise you can be with your statements the better we will benefit from your feedback.

Thank you.